% 1M skyldes lønn til mennesker eller kun compute? (tid = penger hvis det er
% mennesker)

% BF: Har du en referanse vi kan bruke her? Dette er en ganske vanlig quote, men
% husker ikke hvor jeg har sett den, Dette er veldig generelle greier, men vi
% skulle hatt et par gode paper å sitere. 
% LAB: Rob Finn (EBI) bruker denne: https://genomebiology.biomedcentral.com/articles/10.1186/gb-2011-12-8-125
% LAB: Også har vi disse: https://www.veritasgenetics.com/mygenome
In the past decade the generation of biological datasets has been unprecedented,
and the famous "\$1000k genome, \$1M analysis"\cite{} has become more apparent.
% mer om biologi og data 

% Trenger litt tekst som sier at vi skal snakke om stats, presentasjon, dbs,
% microservices og kobling av alle. 

% R og stats 
% LAB: To analyze eller To explore
To analyze the growing biological data sets, there is now a number of analysis
tools in various programming languages.\cite{} In R, there are popular package
repositories such as CRAN \url{cran.r-project.org} or Bioconductor
\url{bioconductor.org} where developers can share software packages and keep
them up-to-date.  These repositories contain software packages for exploring
high-throughput genomic data in one environment. This includes both
pre-processing, e.g. cleaning, removing outliers, and analysing it with known
statistical methods. In other languages such as Python or Go developers can
choose bioinformatics libraries such as BioPython\cite{} and biogo\cite{}
respectively.  To use these packages to their full potential researchers require
a high level of coding skill, and 
% LAB: Omskrive: To fully utilize these packages requires ...coding skills LAB:
% ...og selv da er det ikke alt som enkelt kan løses, som sklaering i R, etc
% LAB: Og hva er problemet og løsningen for dette (vis tools?) 
However all of these software packages require users to have a
high level of coding skills. 

% Vis 
% LAB: Henger ikke sammen med forrige
% LAB: Bedre å starter med denne? Vis tools er løsningen, hvis de finnes for
% problemet. Hvis ikke må sette i gang og kode.
A key part of analysing biological data is to visualize it. Researchers often
start data analysis by using simple visualization techniques to get a quick
overview of the data. Continuing in the data analysis pipeline researchers can
use more advanced visual techniques to explore the data either using
software packages or complete data visualization applications. Traditionally
data visualization applications have been built as desktop applications that
require installation and setup by the user, but now the move is towards software
that run in the web browser without any user setup. 

% WIP WIP WIP
% In systems biology there is a need for visualization tools that can make use
% of the lastest statistical packages and interact with these. It is not enough
% to visualize the results of an analysis pipeline, but interact with the
% different steps from the final visualization. Examples of this interaction can
% be modifying clustering parameters that are used in the analysis.

% Data fra R til vis
To visualize and share results from statistical analyses, the results are often
exported to a data format such as comma-separated values (CSV) and
then visualized using an external tool. This decouples data presentation and
analysis.  
% LAB: Hva er bra/dårlig med decoupling
Through initiatives such as ApacheR and OpenCPU there has been a move
towards embedding scientific computation in data visualization application. This
removes the decoupling of statistical frameworks and interaction with the
analysis results. 

% Koble data sammen med db. 
% LAB: Hva er sammenhengen med programming packages & viz tools?
Interpreting the analysis results require integration of known biology, either
from biological databases such as MSigDB\cite{} or KEGG\cite{}, or through
scientific publications from reference databases such as PubMed\cite{}.
Performing manual lookup into these databases is often tedious and error-prone,
making it necessary to automate the task. Now as more databases provide REST
APIs it is possible to provide software packages for automatic retrival of
database information. In addition, since databases are continuously being
updated, using a REST API to retrieve database information will ensure that the
data is always kept up to date. 

% Hvorfor bryr vi oss om både stats + DBs?  Skal vi si noe om at det blir
% vanligere å bruke REST APIs fremfor å laste ned alt og lese ting lokalt

% Lage apps som gjør alt dette. 

% Hva er micro-services? 
% LAB: Hadde forventet en beskrivelse av ett problem (motivert i paragrafene
% overnfor), som microservices løser
Microservice architectures structures applications into small reusable, loosely
coupled parts. These communicate via lightweight programming language-agnostic
protocols such as HTTP, making it possible to write single applications in
multiple different programming languages. This makes it possible to use the most
suitable programming language for each specific part. E.g. to use R and
Bioconductor to analyse biological data, or C++ and OpenCV for high-performance
computer vision tasks, or HTML, CSS, and JavaScript to build portable
user-interfaces. To build a microservice application, developers bundle each
service in a container that are deployed. Containers are built from
configuration files which describe the operating system, software packages and
versions of these. This makes reproducing an application a trivial task. The
most popular implementation of a software container is Docker\footnote{\url{}},
but others such as Rkt\footnote{\url{}} exist.  

% Har sett et par (bla alex lex sin siste og en VR-artikkel fra i fjor) at det 
% er vanlig med et avsnitt på slutten av bg/intro som beskriver hva artikkelen
% handler om. 
In this paper we describe a novel approach for building data exploration
applications in systems biology. We show that by building applications as a set
of services it is possible to reuse and share its components between
applications. In addition, by packaging the services using container technology
we promote reproducible research. We have used our approach to build a number of
applications, both command-line and web-based. In this paper we show the
viability of our approach by using a web-application, MIxT, for exploring and
comparing transcriptional profiles from blood and tumor samples. 

% LAB: Related bio work? Annen microservice related work? Hvorfor ikke bare bruke de?

\section*{Requirements} 
From our experience building data exploration applications we have identified a
set of reusable services that application developers can use to build a wide
range of applications. The key services of a biological data exploration
application are i) a service for executing for statistical analyses in languages
such as R, and ii) a query service for retrieving meta-data on genes or other
biological entities. On top of these services is possible to build any number
applications and these can be reused by different applications. 

To build these services we need a framework that fulfills the following
requirements: 

\begin{enumerate}
    \item It provides a language-independent approach for integrating, or
        embedding, statistical software, such as R, directly in interactive data
        exploration applications.
    \item It provides an interface to online databases to provide meta-data to
        biological entities. % liker ikke entities-ordet, må ha noe annet 
    \item Its components should be easy to develop, maintain, deploy and share
        between projects. 
\end{enumerate} 


% \subsubsection*{Contributions} 
% Our contributions are: 
% \begin{enumerate}
% % men ikke visualization?
% \item An approach for developing data exploration applications for systems
% biology that combine statistical analyses with online databases.  
% \item A demonstration of its viability through N different applications. 
% \item Performance evaluation of its central data analysis component. 
% \end{enumerate} 

