In recent years the biological community has  generated  an
unprecedented ammount of data. While the cost of data collection has
drastically decreased, data analysis continue to be a larger fraction of the
total experiment cost.\cite{sboner2011real}  An important part of data analysis
includes the time spent by human experts interpreting the results. This calls
for novel methods for building data analysis tools for data exploration and
interpretation. 

In the field of systems biology, data exploration applications need to link
results to relevant prior knowledge. In the later years there has been a
tremendous effort to curate databases with relevant information on genes and
processes. Databases such as the Molecular Signatures Database
(MSigDB)\footnote{\url{software.broadinstitute.org/gsea/msigdb}} and the Kyoto
Encyclopedia of Genes and Genomes (KEGG)\footnote{\url{kegg.jp}} both provide
interfaces to retrieve data that can be used to better the understanding of
data analysis results. 

Several tools for biological data analysis are now available in various
programming languages. These include a wide variety of bioinformatics
methodologies and graphical analysis tools.
In the R statistical programming language developers share software through
repositories such as CRAN\footnote{\url{cran.r-project.org}} or
Bioconductor\footnote{\url{bioconductor.org}}.  In other languages, libraries
for biological bomputation are often availalbe like BioPython\cite{biopython}
and biogo\cite{biogo} for Python or Go, respectively. Projects such as
Galaxy\footnote{\url{galaxyproject.org}} and Common Workflow Language
(CWL)\footnote{\url{commonwl.org}} enable resarchers to build and run
biological data analysis pipelines consisting of a wide range of tools.
Although these framework are tremendously helpful, we need novel approaches to
build applications that integrate high-performance bioinformatics tools with 
specialized user-interfaces and interactive visualizations

Different programming languages solve different tasks.  For example, new
biological data analysis techniques are quickly realeased in R and its package
repositories; high-performance computer vision tasks are performed in C++ and
OpenCV; and portable user interfaces more easily built in HTML, CSS and
JavaScript.  While it would make development easy, we argue that data
exploration applications in systems biology should not be written in a single
programming language. Therefore applications that integrate novel statistical
analysis tools, interactive visualizations, and biological databases likely
need to include several components written in different languages. 

A microservice architecture structures an application into small reusable,
loosely coupled parts. These communicate via lightweight programming
language-agnostic protocols such as HTTP, making it possible to write single
applications in multiple different programming languages. This way the
most suitable programming language is used for each specific part. To build a
microservice application, developers bundle each service in a container.
Containers are built from configuration files which describe the operating
system, software packages and versions of these. 
This makes reproducing the analyses, database lookups, library versions in an
application a trivial task. The most popular implementation of a software
container is Docker\footnote{\url{docker.com}}, but others such as
Rkt\footnote{\url{coreos.com/rkt}} exist. Initiatives such as
BioContainers\footnote{\url{biocontainers.pro}} now provide containers
pre-installed with different bioinformatics tools. 

From our experience we have identified a set of components and features we
need to build data exploration applications.

\begin{enumerate}
    \item A low-latency language-independent approach for integrating, or
        embedding, statistical software, such as R, directly in an application. 
    \item Low latency language-independent interface to online reference
        databases in biology that users can query to better understand resulst
        from statistical analyses. 
    \item A simple method for deploying and sharing the components of an
        application between projects. 
\end{enumerate} 


In this paper, we describe a novel approach for building data exploration
applications in systems biology. We show that by building applications as a set
of services we can reuse and share its components between applications. 
The key services of a biological data exploration application are i) a compute
service for executing statistical analyses in languages such as R, ii) a
database query service for retrieving information from biological databases, and
iii) the user-facing visualizations and user-interfaces. 
In addition, by packaging the services using container technology they are easy
to deploy, simple to reproduce, and easy to share between projects. 
We have
used our approach to build a number of applications, both command-line and
web-based. In this paper we describe how we used our approach to develop MIxT,
a web application for exploring and comparing transcriptional profiles from
blood and tumor samples. 
