% Hva er korrekt: "in" system biology eller "for" system biology?
In this section we aim to cover the existing approaches for building interactive
data exploration applications in systems biology. 

% TODO: Add related work where somebody has built a system for exploring some
% data where 

% Det er en utfordring å ha related work tidlig i dette paperet. Slik det er nå,
% er det ikke sikkert leseren skjønner hvorfor diss systemene diskuteres, hva
% som er sammenhengen mellom de, og hva det har å gjrøe med Kvik.

\subsubsection*{OpenCPU} 
OpenCPU is a system for embedded scientific computing and reproducible
research.\cite{opencpu} It provides an HTTP API to the R programming language to
provide an interface with statistical methods. It enables users to make function
calls to any R package and retrieve the results in a wide variety of formats
such as json or pdf. 
Users can chose to host their own R server or use public servers, and OpenCPU
works in a single-user setting within an R session, or a multi-user setting
facilitating multiple parallel requests. 
OpenCPU provides a Javascript library for interfacing with R, as well as Docker
containers for easy installation. OpenCPU has been used to build multiple
applications.\footnote{\url{opencpu.org/apps.html}}.
% Design patterns er ikke nevnt enda
In Kvik we provide a package
to interface with OpenCPU servers from the Go programming language since it
follows the design pattern we have chosen to interface with data and analyses. 


\subsubsection*{Renjin} 
% Skjønner ikke poenget med Renjin. Dvs hva fordelen med å kjøre R kode på en
% JVM er.
Renjin is a JVM-based interpreter for the R programming language.\cite{renjin}
It targets developers who want to integrate the R interpreter in web
applications. Since it is built on top of the JVM it allows developers to write
data exploration applications that interact directly with R code. Although
Renjin supports a large number of CRAN packages it cannot access any R package
(e.g. from BioConductor) without modification. This makes the programming effort
to use Renjin as an interface with R higher. 

\subsubsection*{Shiny} 
Shiny is a web application framework for R.\cite{shiny} It allows developers to
build web applications in R without having to have any knowledge about HTML, CSS
or Javascript. It provides a widget library to provide more advanced Javascript
visualizations such as Leaflet for maps or threejs for WebGL-accellerated
graphics. Developers can choose to host their own web server with the user-built
Shiny Apps, or host them on public servers. Shiny forces users to implement data
exploration applications in R, limiting the functionality to the 
widgets and libraries in Shiny. 

% Synes SparkR burde nevnes

\subsubsection*{Biogo} 
Biogo is a bioinformatics library in Go. It provides functionality to analyse
genomic and metagenomc datasets in the go programming
language.\cite{Kortschak005033} Using the go programming language the developers
are able to provide high-performance parallel processing in a clean and simple
programming language. 

\subsubsection*{Cytoscape} 
Cytoscape is an open source software platform for visualizing complex
networks and integrating these with any type of attribute
data\cite{shannon2003cytoscape}. It allows for analysis and visualization in the
same platform. Users can add additional features, such as databases connections
or new layouts, through Apps. To bring the visualization and analysis
capabilities to the web the creators of Cytoscape have developed
Cytoscape.js\footnote{\url{js.cytoscapejs.org}}, a Javascript library to create
interactive graph visualizations. 

cyREST. 



