% LAB: Si heller hvilke områder som er beskrevet nedenfor
In this section we aim to cover some of the existing systems for building
interactive data exploration applications in systems biology. 

% LAB: Integrate med hva?
\subsection*{Integrate Statistical Analyses} 
% LAB: Siden Kvik nå er beskrevet kan teksten ta hensyn til det. Så feks: It offers -> Similary to Kvik, it offers 
OpenCPU is a system for embedded scientific computing and reproducible
research.\cite{opencpu} It offers an HTTP API to the R programming language to
provide an interface with statistical methods. It enables users to make function
calls to any R package and retrieve the results in a wide variety of formats
such as json or pdf. 
Users can chose to host their own R server or use public servers, and OpenCPU
works in a single-user setting within an R session, or a multi-user setting
facilitating multiple parallel requests. This makes OpenCPU suitable
for building a service that can run statistical analyses. 
OpenCPU provides a Javascript library for interfacing with R, as well as Docker
containers for easy installation and OpenCPU has been used to build multiple
applications.\footnote{\url{opencpu.org/apps.html}}.
% LAB: også er det viktig å avslutte med hvorfor vi ikke like godt brukte OpenCPU eller de andre tingene som beskrives her.

Renjin is a JVM-based interpreter for the R programming language.\cite{renjin}
It targets developers who want to integrate the R interpreter in web
applications. Since it is built on top of the JVM it allows developers to write
data exploration applications that interact directly with R code, both runnin on
top of the JVM. Although Renjin supports a large number of CRAN packages it
cannot access any R package (i.e.  any package from BioConductor or CRAN)
without modification. This makes the programming effort to use Renjin as an
interface to R higher. 

Shiny is a web application framework for R\footnote{\url{shiny.rstudio.com}}
It allows developers to
build web applications in R without having to have any knowledge about HTML, CSS
or Javascript. Its widget library to provides more advanced Javascript
visualizations such as Leaflet\footnote{\url{leafletjs.com}} for maps or
three.js\footnote{\url{threejs.org}} for WebGL-accellerated
graphics. Developers can choose to host their own web server with the user-built
Shiny Apps, or host them on public servers. Shiny forces users to implement data
exploration applications in R, limiting the functionality to the 
widgets and libraries in Shiny. 

% Synes SparkR burde nevnes

\subsubsection*{Biogo} 
\emph{this one is a bit out of place.}

Biogo is a bioinformatics library written in Go. It provides functionality to
analyze genomic and metagenomic datasets in the go programming
language.\cite{Kortschak005033} Using the go programming language the developers
are able to provide high-performance parallel processing in a clean and simple
programming language. 

\subsection*{Visualization frameworks} 
Cytoscape is an open source software platform for visualizing complex
networks and integrating these with any type of attribute
data\cite{shannon2003cytoscape}. It allows for analysis and visualization in the
same platform. Users can add additional features, such as databases connections
or new layouts, through Apps. One such app is cyREST which allows external network
creation and analysis through a REST API\cite{ono2015cyrest}.
To bring the visualization and analysis
capabilities to the web the creators of Cytoscape have developed
Cytoscape.js\footnote{\url{js.cytoscapejs.org}}, a Javascript library to create
interactive graph visualizations. 


Caleydo is a framework for building applications for visualizing and exploring
biomolecular data\cite{cleydo}. Until 2014 it was a standalone tool that needed
to be downloaded, but the Caleydo team are now making the tools web-based. There
have been several applications built using Caleydo: StratomeX for exploring
stratified heterogeneous datasets for disease subtype analysis\cite{stratomex};
Pathfinder for exploring paths in large multivariate graphs\cite{pathfinder};
UpSet to visualize and analyse sets, their intersections and
aggregates\cite{upset}; Entourage and enRoute to explore and visualize
biological pathways \cite{entourage}\cite{enroute}; LineUp to explore rankings
of items based on a set of attributes\cite{lineup}; and Domino for exploring
subsets across multiple tabular datasets\cite{domino}. 

BioJS is an open-source JavaScript framework for biological data
visualization.\cite{gomez2013biojs} It provides a community-driven online
repository with a wide range components for visualizing biological data
contributed by the bioinformatics community. BioJS builds on
node.js\footnote{\url{nodejs.org}} providing both server-side and client-side
libraries. 


\subsection*{WIP: Biological Databases} 
Maybe some words here on how to get data out of the different biological
databases? 

\subsection*{WIP: Microservices, Docker etc.} 
... 

\subsection*{Kvik and Kvik Pathwys}
We have previously built a system for interactively exploring gene expression
data in context of  biological pathways.\cite{fjukstad2015kvik} Kvik Pathways is
a web application that integrates gene expression data from the Norwegian Women
and Cancer (NOWAC) cohort together with pathway images from the Kyoto
Encyclopedia of Genes and Genomes (KEGG). We used the experience building Kvik
Pathways to completely re-design and re-implement
the R interface in Kvik. From having an R server that can run a set of functions
from an R script, it now has a clean interface to call any function from any R
package, not just retrieving data as a text string but in a wide range of
formats. We also re-built the database interface, now moving it into its own
service. This makes it possible to leverage its caching capabilities to improve
latency. 

